pkg/crow/samples.go

```diff
@@ -0,0 +1,70 @@
+package crow
+
+import (
+	"fmt"
+	"math"
+	"math/rand"
+	"time"
+
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+type sample struct {
+	ScrapeTime time.Time
+	Labels     prometheus.Labels
+	Value      float64
+
+	// How many times this sample has attempted to be valdated. Starts at 0.
+	ValidationAttempt int
+}
+
+// Ready checks if this sample is ready to be validated.
+func (s *sample) Ready(now time.Time) bool {
+	// Exponential backoff from 500ms up (500ms * 2^attempt).
+	backoff := (500 * time.Millisecond) * time.Duration(math.Pow(2, float64(s.ValidationAttempt)))
+	return now.After(s.ScrapeTime.Add(backoff))
+}
+
+type sampleGenerator struct {
+	numSamples int
+	sendCh     chan<- []*sample
+	r          *rand.Rand
+}
+
+const validationSampleName = "crow_validation_sample"
+
+func (sg *sampleGenerator) Describe(ch chan<- *prometheus.Desc) {
+	ch <- prometheus.NewDesc(
+		validationSampleName, "Sample to validate",
+		[]string{"sample_num"},
+		prometheus.Labels{},
+	)
+}
+
+func (sg *sampleGenerator) Collect(ch chan<- prometheus.Metric) {
+	var (
+		scrapeTime = time.Now()
+		desc       = prometheus.NewDesc(
+			validationSampleName, "Sample to validate",
+			[]string{"sample_num"},
+			prometheus.Labels{},
+		)
+	)
+
+	samples := make([]*sample, sg.numSamples)
+	for s := 0; s < sg.numSamples; s++ {
+		samples[s] = &sample{
+			ScrapeTime: scrapeTime,
+			Labels:     prometheus.Labels{},
+			Value:      float64(sg.r.Int63n(1_000_000)),
+		}
+
+		ch <- prometheus.MustNewConstMetric(
+			desc,
+			prometheus.GaugeValue,
+			samples[s].Value, fmt.Sprintf("sample_%d", s),
```

Comment:
I don't think generating the same label each time is going to cut it here. Since series will stay alive for 4 minutes, it's pretty unlikely that we'd ever see a "missing" error after the first scrape. We'd just see a "mismatch" error since it'd be validating the sample from the previous scrape. 

Maybe generate a random label each time? There's not a ton of metrics so a churn of ~10 series per 5-60 seconds isn't a huge deal.